62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+
+        const double maxMajorEnds = ends.top();
+
+		const Segment nextStartEvent = starts.empty() ? Segment() : starts.top();
+		const double minMajorStart = nextStartEvent.originalBezier ? nextStartEvent.originalBezier->evaluateBezier(nextStartEvent.t_start)[major] : 0.0;
+
+        // We check which event, within start, end and intersection events have the smallest
+        // major coordinate at this point
+
+		auto intersectionVisit = [&]()
+		{
+			auto newMajor = intersections.top();
+			if (debugOutput && isCurrentDebugStep)
+				drawDebugLine(float64_t2(-1000.0, newMajor), float64_t2(1000.0, newMajor), float32_t4(0.0, 0.0, 0.8, 1.0));
+			intersections.pop(); // O(n)
+			//std::cout << "Intersection event at " << newMajor << "\n";
+			return newMajor;
+		};
+
+        // next start event is before next end event
+        if (nextStartEvent.originalBezier && minMajorStart < maxMajorEnds)
+        {
+            // next start event is before next intersection event
+            // (start event)
+            if (intersections.empty() || minMajorStart < intersections.top()) // priority queue top() is O(1)
+            {
+                starts.pop();
+                newMajor = minMajorStart;
+				addStartSegmentToCandidates = true;
+				if (debugOutput && isCurrentDebugStep)
+                    drawDebugLine(float64_t2(-1000.0, newMajor), float64_t2(1000.0, newMajor), float32_t4(0.0, 0.8, 0.0, 1.0));
+                //std::cout << "Start event at " << newMajor << "\n";
+            }
+            // (intersection event)
+            else newMajor = intersectionVisit();
+        }
+        // next end event is before next intersection event
+        // (intersection event)
+        else if (!intersections.empty() && intersections.top() < maxMajorEnds)
+            newMajor = intersectionVisit();
+        // (end event)
+        else
+        {
+            newMajor = maxMajorEnds;
+            ends.pop();
+            if (debugOutput && isCurrentDebugStep)
+                drawDebugLine(float64_t2(-1000.0, newMajor), float64_t2(1000.0, newMajor), float32_t4(0.0, 0.0, 0.8, 1.0));
+            //std::cout << "End event at " << newMajor << "\n";
+        }
+        // spawn quads for the previous iterations if we advanced
+        //std::cout << "New major: " << newMajor << " Last major: " << lastMajor << "\n";
+        if (newMajor > lastMajor)
+        {
+			if (debugOutput && isCurrentDebugStep)
+				drawDebugLine(float64_t2(-1000.0, lastMajor), float64_t2(1000.0, lastMajor), float32_t4(0.1, 0.1, 0.0, 0.5));
+            // trim
+            const auto candidatesSize = std::distance(activeCandidates.begin(),activeCandidates.end());
+			//std::cout << "Candidates size: " << candidatesSize << "\n";
+            // because n4ce works on loops, this must be true
+            assert((candidatesSize % 2u)==0u);
+            for (auto i=0u; i< candidatesSize;)
+            {
+                auto left = activeCandidates[i++];
```

Comment:
I don't suggest using auto for basic and simple types, it makes it slightly harder to read and makes me want to hover on it.
and I think you rely on this being a copy because you're modifying the `ends` later on and you don't want to change the original ends.

I suggest you use `splitCurveRange(*originalBezier, t_start, newEnd);` directly and make this a `const Segment&`